src/main/java/pl/com/itsystems/openweathermap/WeatherService.java

```diff
@@ -0,0 +1,34 @@
+package pl.com.itsystems.openweathermap;
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+import pl.com.itsystems.openweathermap.dto.WeatherResponseDto;
+import pl.com.itsystems.openweathermap.exception.CityNotFoundException;
+
+@Service
+public class WeatherService {
+
+    @Value("${appid}")
+    private String apiKey;
+
+    public WeatherResponseDto getWeather(String city, String units) {
+        RestTemplate restTemplate = new RestTemplate();
+        try {
+            String url = UriComponentsBuilder.fromHttpUrl("https://api.openweathermap.org")
+                    .path("/data/2.5/weather")
+                    .queryParam("q", city)
+                    .queryParam("appid", apiKey)
+                    .queryParam("units", units)
+                    .queryParam("lang", "pl")
+                    .toUriString();
+            return restTemplate.getForObject(url, WeatherResponseDto.class);
+        } catch (HttpClientErrorException e) {
+            throw new CityNotFoundException(city);
+        } catch (Exception e) {
```

Comment:
To jest niestety zbyt ogólne. Jeśli podam niepoprawny klucz api to również zostanie rzucony HttpClientErrorException. Powód można odczytać z treści:

`{"cod":401, "message": "Invalid API key. Please see https://openweathermap.org/faq#error401 for more info."}`

Taki "piękny" kawałek kodu może dać pomysł:

```java
            final String responseBodyAs = e.getResponseBodyAs(String.class);
            ObjectMapper objectMapper = new ObjectMapper();
			final ErrorDto errorDto;
			try {
				errorDto = objectMapper.readValue(responseBodyAs, ErrorDto.class);
			} catch (JsonProcessingException ex) {
				throw new RuntimeException(ex);
			}
			if (errorDto.cod == 404) {
                throw new CityNotFoundException(city);
            }
            throw new RuntimeException(e);
```

gdzie ErrorDto to:

```java
record ErrorDto(int cod, String message) {}
```